client/src/App.tsx

```diff
@@ -1,67 +1,262 @@
+import React, { useEffect, useState } from 'react';
+import * as elliptic from 'elliptic';
 
-const API_URL = "http://localhost:8080";
+const API_URL = 'http://localhost:8080';
+
+const ec = new elliptic.ec('secp256k1');
 
 function App() {
-  const [data, setData] = useState<string>();
+  const [data, setData] = useState<string>('');
+  const [isTokenVisible, setIsTokenVisible] = useState<boolean>(true);
+  const [isTokenBtnVisible, setIsTokenBtnVisible] = useState<boolean>(true);
+  const [accessToken, setAccessToken] = useState<string>('');
+  const [isTampered, setTampered] = useState<boolean>(false);
+
+  console.log({ accessToken });
 
   useEffect(() => {
     getData();
   }, []);
 
+  useEffect(() => {
+    if (!isTokenVisible) return;
+
+    const timeoutId = setTimeout(() => {
+      setIsTokenVisible(false);
+    }, 30000);
+
+    return () => clearTimeout(timeoutId);
+  }, [isTokenVisible]);
+
   const getData = async () => {
     const response = await fetch(API_URL);
-    const { data } = await response.json();
+    const data = await response.json();
     setData(data);
   };
 
   const updateData = async () => {
+    const keyPair = ec.keyFromPrivate(accessToken, 'hex');
+
+    const signature = keyPair.sign(data).toDER('hex');
+
     await fetch(API_URL, {
-      method: "POST",
-      body: JSON.stringify({ data }),
+      method: 'POST',
+      body: JSON.stringify({ data, signature }),
       headers: {
-        Accept: "application/json",
-        "Content-Type": "application/json",
+        Accept: 'application/json',
+        'Content-Type': 'application/json',
       },
     });
 
     await getData();
   };
 
   const verifyData = async () => {
-    throw new Error("Not implemented");
+    console.log('verifyData');
+    try {
+      const data = await fetch(`${API_URL}/verify`, { method: 'GET' });
+      const { data: resData } = await data.json();
+
+      console.log(resData[0].data, resData[0].signature, resData[0].key);
+
+      const keyPair = ec.keyFromPrivate(accessToken, 'hex');
+
+      console.log(keyPair);
+      const isVerified = ec.verify(
+        resData[0].data,
+        resData[0].signature,
+        resData[0].key,
+        'hex'
+      );
+
+      console.log(isVerified);
+
+      setTampered(!isVerified);
+    } catch (error) {
+      setTampered(true);
+      console.log(error);
+    }
+  };
+
+  const tamperDataToCheck = async () => {
+    const keyPair = ec.keyFromPrivate(accessToken, 'hex');
+
+    const signature = keyPair.sign(data).toDER('hex');
+
+    await fetch(API_URL, {
+      method: 'POST',
+      body: JSON.stringify({
+        data: data + ' tampered',
+        signature: signature + 'tampered',
+      }),
+      headers: {
+        Accept: 'application/json',
+        'Content-Type': 'application/json',
+      },
+    });
+
+    await getData();
+  };
+
+  const getAccessToken = async () => {
+    try {
+      const response = await fetch(`${API_URL}/access-token`, {
```

Comment:
The private key used for signatures is generated by the server. Is it better to generate and save it on the client side?

Reply:
You are right. It is better to generate the keys on client.